### PR TITLE
added validators

### PR DESCRIPTION
**Pull request by Yugue**
*Added validators for inputs and dropdowns of release initialization step.*

Only when all the parameters are saved properly, `Continue` button becomes enabled. These are the validators' criteria:

- The dropdowns must have values.
- All inputs have regex to be satisfied (The regex are tested separately).
- cherrypicks and dart revisions can be empty strings.
- cherrypicks are git hashes delimited by a comma, trailing comma not allowed.
- dart revision can only be one git hash.

Screen recording:
https://user-images.githubusercontent.com/20194490/139352685-1f037689-c945-4066-b11d-75b869add456.mov

*[Issue](https://github.com/flutter/flutter/issues/91925)*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: release_dashboard/lib/logic/git_regex.dart [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+/// Supports multiple git hashes delimited by a comma or an empty string.
+///
+/// valid: c7714158950347,c7714158950347,c7714158950347
+/// valid: c7714158950347
+/// unvalid:   c7714158950347,@@cccccc
+/// unvalid (cannot end with a comma):   c7714158950347,c7714158950347,
+var multiGitHashPattern = [r'^$|^', gitHashPattern, r'(?:,', gitHashPattern, r')*$'].join();
+
+final RegExp multiGitHashRegex = RegExp(multiGitHashPattern);
```
Comment (christopherfujino) [inline release_dashboard/lib/logic/git_regex.dart:18]:
```
final RegExp multiGitHashRegex = RegExp(r'^([0-9a-z]{40}(,[0-9a-z]{40})*)?$');
```

File: release_dashboard/lib/logic/git_regex.dart [hunk 2]
```diff
@@ -0,0 +1,20 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+const String gitHashPattern = r'[0-9a-f]{40}';
+
+// Allows empty string because the dart revision which is the only input using this regex could be empty
+final RegExp gitHashRegex = RegExp([r'^$|^', gitHashPattern, r'$'].join());
```
Comment (christopherfujino) [inline release_dashboard/lib/logic/git_regex.dart:8]:
```
final RegExp gitHashRegex = RegExp(r'^([0-9a-f]{40})?$');
```

If this is only used for `dartRevision` then maybe name it accordingly, too.

File: release_dashboard/lib/logic/git_regex.dart [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+///
+/// valid: c7714158950347,c7714158950347,c7714158950347
+/// valid: c7714158950347
+/// unvalid:   c7714158950347,@@cccccc
+/// unvalid (cannot end with a comma):   c7714158950347,c7714158950347,
+var multiGitHashPattern = [r'^$|^', gitHashPattern, r'(?:,', gitHashPattern, r')*$'].join();
+
+final RegExp multiGitHashRegex = RegExp(multiGitHashPattern);
+
+final RegExp candidateBranchRegex = RegExp(r'^flutter-(\d+)\.(\d+)-candidate\.(\d+)$');
```
Comment (christopherfujino) [inline release_dashboard/lib/logic/git_regex.dart:20]: make sure you're calling `.trim()` on all inputs you compare against these regexes, since you're using `^` and `$`

File: release_dashboard/lib/widgets/create_release_substeps.dart [hunk 4]
```diff
@@ -112,30 +140,67 @@ class CreateReleaseSubstepsState extends State<CreateReleaseSubsteps> {
...
+      case 6:
+        formRegexValidator = gitHashRegex;
+        break;
+      default:
+        break;
+    }
+
     return TextFormField(
       key: Key(CreateReleaseSubsteps.substepTitles[index]),
+      autovalidateMode: AutovalidateMode.always,
```
Comment (christopherfujino) [inline release_dashboard/lib/widgets/create_release_substeps.dart:185]: I think `AutovalidateMode.onUserInteraction` would probably be better

File: release_dashboard/lib/widgets/create_release_substeps.dart [hunk 5]
```diff
@@ -112,30 +140,67 @@ class CreateReleaseSubstepsState extends State<CreateReleaseSubsteps> {
...
   final int index;
   final SetReleaseData setReleaseData;
   final String? hintText;
+  final ChangeIsEachInputValid changeIsInputValid;
 
   @override
   Widget build(BuildContext context) {
+    late RegExp formRegexValidator;
+    // assigns corresponding regex to each input
+    switch (index) {
```
Comment (christopherfujino) [inline release_dashboard/lib/widgets/create_release_substeps.dart:164]: I think rather than hard-coding into this widget the behavior and order of the different fields, it would probably be better to create different classes for the different types of input.

File: release_dashboard/lib/widgets/create_release_substeps.dart [hunk 6]
```diff
@@ -112,30 +140,67 @@ class CreateReleaseSubstepsState extends State<CreateReleaseSubsteps> {
...
         setReleaseData(CreateReleaseSubsteps.substepTitles[index], data);
+        if (!formRegexValidator.hasMatch(data)) {
+          changeIsInputValid(index, false);
+        } else {
+          changeIsInputValid(index, true);
+        }
+      },
+      validator: (String? value) {
+        if (!formRegexValidator.hasMatch(value ?? '')) {
+          return '${CreateReleaseSubsteps.substepTitles[index].replaceAll(' (if necessary)', '')} not in a valid format!';
```
Comment (christopherfujino) [inline release_dashboard/lib/widgets/create_release_substeps.dart:200]: Isn't this value already accessible in scope as `value`?

Comment (christopherfujino) [reply release_dashboard/lib/widgets/create_release_substeps.dart:200]: Also, once you refactor this into multiple classes for the different types of input, I think it would be nice to hardcode in an example of the valid input. Something like, `"This must be a valid candidate branch string, e.g. 'flutter-1.2-candidate.3'"`

File: release_dashboard/lib/widgets/create_release_substeps.dart [hunk 7]
```diff
@@ -47,6 +60,13 @@ class CreateReleaseSubstepsState extends State<CreateReleaseSubsteps> {
     });
   }
 
+  // When isEachInputValid[0] is true, the first parameter is valid. If it false, it is unvalid.
```
Comment (christopherfujino) [inline release_dashboard/lib/widgets/create_release_substeps.dart:63]: nit, in English the word is "invalid"
```suggestion
  // When isEachInputValid[0] is true, the first parameter is valid. If it false, it is invalid.
```

File: release_dashboard/lib/widgets/create_release_substeps.dart [hunk 8]
```diff
@@ -56,53 +76,61 @@ class CreateReleaseSubstepsState extends State<CreateReleaseSubsteps> {
...
         const SizedBox(height: 20.0),
         Center(
-          // TODO(Yugue): Add regex validation for each parameter input
-          // before Continue button is enabled, https://github.com/flutter/flutter/issues/91925.
           child: ElevatedButton(
             key: const Key('step1continue'),
-            onPressed: () {
-              widget.nextStep();
-            },
+            onPressed: isEachInputValid.contains(false) // Continue button is disabled if any of the inputs are invalid
```
Comment (christopherfujino) [inline release_dashboard/lib/widgets/create_release_substeps.dart:129]: nit: rather than

```
() {
  widget.nextStep(0;
},
```

You can do a tear-off: `onPressed: isEachInputValid.contains(false) ? null : widget.nextStep,`

File: release_dashboard/lib/widgets/create_release_substeps.dart (hunk 5 above)
Comment (christopherfujino) [reply release_dashboard/lib/widgets/create_release_substeps.dart:164]: On a higher level, as part of encapsulation, a widget probably should not need to know its position relative to its siblings.

File: release_dashboard/test/logic/git_regex_test.dart [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+    test('Multi Git hash pattern rejects if there is an invalid hash', () {
+      expect(multiGitHashRegex.hasMatch(invalidMultiHash2), false);
+    });
+
+    test('Multi Git hash pattern rejects a single invalid hash', () {
+      expect(multiGitHashRegex.hasMatch(invalidMultiHash3), false);
+    });
+  });
+
```
Comment (christopherfujino) [inline release_dashboard/test/logic/git_regex_test.dart:69]: Well done, these are good unit tests!

File: release_dashboard/test/widgets/stepper_test.dart [hunk 10]
```diff
@@ -34,6 +34,6 @@ void main() {
     await tester.tap(find.text('Initialize a New Flutter Release'));
     await tester.pumpAndSettle();
 
-    expect(tester.widget<Stepper>(find.byType(Stepper)).currentStep, equals(1));
+    expect(tester.widget<Stepper>(find.byType(Stepper)).currentStep, equals(0));
```
[the remark this answers is not in the document]
Comment (christopherfujino) [reply release_dashboard/test/widgets/stepper_test.dart:35]: Does this still test what the description says? I'm a little confused as to what it's testing.

File: release_dashboard/lib/logic/git_regex.dart (hunk 3 above)
Comment (christopherfujino) [reply release_dashboard/lib/logic/git_regex.dart:16]: Why not?

Comment (CaseyHillers) [reply release_dashboard/lib/logic/git_regex.dart:16]: +1 - trim() is pretty quick to call, and improves the UX. It's common to do input sanitization for common user issues like this

File: release_dashboard/test/widgets/stepper_test.dart (hunk 10 above)
[the remark this answers is not in the document]
Comment (christopherfujino) [reply release_dashboard/test/widgets/stepper_test.dart:35]: Can we delete the test, then?

File: release_dashboard/lib/widgets/create_release_substeps.dart (hunk 5 above)
Comment (christopherfujino) [reply release_dashboard/lib/widgets/create_release_substeps.dart:160]: Yes, create a different widget for each different "type" of input.

File: release_dashboard/lib/widgets/create_release_substeps.dart (hunk 6 above)
Comment (christopherfujino) [reply release_dashboard/lib/widgets/create_release_substeps.dart:196]: > Response from Alex: `CreateReleaseSubsteps.substepTitles[index]` is the title of the input element, and `value` is the input the user provides.

Ahh, I see, you're right.

File: release_dashboard/lib/widgets/create_release_substeps.dart [hunk 11]
```diff
@@ -36,6 +38,17 @@ class CreateReleaseSubstepsState extends State<CreateReleaseSubsteps> {
   // Initialize a public state so it could be accessed in the test file.
   @visibleForTesting
   late Map<String, String?> releaseData = <String, String?>{};
+  @visibleForTesting
+  List<bool> isEachInputValid = List<bool>.filled(CreateReleaseSubsteps.substepTitles.length, false);
```
Comment (christopherfujino) [inline release_dashboard/lib/widgets/create_release_substeps.dart:42]: what about using a map here to store the bools? an array where the indices are semantically meaningful makes the code harder to read and refactoring more difficult.

File: release_dashboard/lib/logic/git.dart [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+final String multiGitHashErrorMsg =
+    'Must be one or more groups of 40 alphanumeric characters delimited by a comma or an empty string.';
+
+final RegExp candidateBranchRegex = RegExp(r'^flutter-(\d+)\.(\d+)-candidate\.(\d+)$');
+final String candidateBranchErrorMsg = "Must be a valid candidate branch string, e.g. 'flutter-1.2-candidate.3'";
+
+final String githubRemoteErrorMsg = "Must be a valid Github remote string, e.g. 'git@github.com:user/flutter.git'";
+
+/// Class that accepts only the index parameter as an identifier to the type of input.
+class git {
```
Comment (CaseyHillers) [inline release_dashboard/lib/logic/git.dart:27]: `git` is very generic. I think a more specific name would be `GitCommit` or `GitHash`

https://dart.dev/guides/language/effective-dart/style#do-name-types-using-uppercamelcase

File: release_dashboard/lib/logic/git.dart [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+/// invalid:   c7714158950347,@@cccccc
+/// invalid (cannot end with a comma):   c7714158950347,c7714158950347,
+var multiGitHashRegex = RegExp(r'^$|^[0-9a-z]{40}(?:,[0-9a-z]{40})*$');
+final String multiGitHashErrorMsg =
+    'Must be one or more groups of 40 alphanumeric characters delimited by a comma or an empty string.';
+
+final RegExp candidateBranchRegex = RegExp(r'^flutter-(\d+)\.(\d+)-candidate\.(\d+)$');
+final String candidateBranchErrorMsg = "Must be a valid candidate branch string, e.g. 'flutter-1.2-candidate.3'";
+
+final String githubRemoteErrorMsg = "Must be a valid Github remote string, e.g. 'git@github.com:user/flutter.git'";
```
Comment (CaseyHillers) [inline release_dashboard/lib/logic/git.dart:24]: These will all become globals once imported. Prefer storing them in the class

https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#minimize-the-visibility-scope-of-constants

Prefer prefixing constants with a `k`

https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#begin-global-constant-names-with-prefix-k

File: release_dashboard/lib/logic/git.dart [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+  git({required this.name});
+
+  final String name;
+
+  /// Return the regex and the error message if the regex fails in a map for each type of input.
+  Map<String, Object> getRegexAndErrorMsg() {
+    switch (name) {
+      case 'Candidate Branch':
+        return <String, Object>{'regex': candidateBranchRegex, 'errorMsg': candidateBranchErrorMsg};
+      case 'Framework Mirror':
```
Comment (CaseyHillers) [inline release_dashboard/lib/logic/git.dart:37]: The should all be separate classes, like `GitRemote`, `GitBranch`, etc

File: release_dashboard/lib/widgets/create_release_substeps.dart [hunk 15]
```diff
@@ -2,6 +2,7 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import '../logic/git.dart';
```
Comment (CaseyHillers) [inline release_dashboard/lib/widgets/create_release_substeps.dart:5]: this import is out of order.

Add pub.dev/packages/flutter_lints to catch these issues

File: release_dashboard/lib/widgets/create_release_substeps.dart [hunk 16]
```diff
@@ -33,9 +34,35 @@ class CreateReleaseSubsteps extends StatefulWidget {
 }
 
 class CreateReleaseSubstepsState extends State<CreateReleaseSubsteps> {
-  // Initialize a public state so it could be accessed in the test file.
+  // Initialize a public state so it could be accessed in the test file
```
Comment (CaseyHillers) [inline release_dashboard/lib/widgets/create_release_substeps.dart:37]:
```suggestion
  /// Initialize a public state so it could be accessed in the test file
```

File: release_dashboard/lib/widgets/create_release_substeps.dart [hunk 17]
```diff
@@ -33,9 +34,35 @@ class CreateReleaseSubsteps extends StatefulWidget {
...
   @visibleForTesting
   late Map<String, String?> releaseData = <String, String?>{};
+  // When isEachInputValid['substep1'] is true, 'substep1' is valid. If it false, it is invalid
```
Comment (CaseyHillers) [inline release_dashboard/lib/widgets/create_release_substeps.dart:40]:
```suggestion
  /// When [substep] in [isEachInputValid] is true, [substep] is valid. Otherwise, it is invalid
```

File: release_dashboard/lib/logic/git.dart [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+final RegExp gitHashRegex = RegExp(r'^$|^[0-9a-f]{40}$');
+final String gitHashErrorMsg = 'Must be 40 alphanumeric characters or an empty string.';
+
+/// Supports multiple git hashes delimited by a comma or an empty string.
+///
+/// valid: c7714158950347,c7714158950347,c7714158950347
+/// valid: c7714158950347
+/// invalid:   c7714158950347,@@cccccc
+/// invalid (cannot end with a comma):   c7714158950347,c7714158950347,
+var multiGitHashRegex = RegExp(r'^$|^[0-9a-z]{40}(?:,[0-9a-z]{40})*$');
```
Comment (CaseyHillers) [inline release_dashboard/lib/logic/git.dart:17]: Avoid var

https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-var-and-dynamic

File: release_dashboard/lib/logic/git.dart [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:conductor_core/conductor_core.dart';
+
+// Allows empty string because the dart revision which is the only input using this regex could be empty
+final RegExp gitHashRegex = RegExp(r'^$|^[0-9a-f]{40}$');
+final String gitHashErrorMsg = 'Must be 40 alphanumeric characters or an empty string.';
+
+/// Supports multiple git hashes delimited by a comma or an empty string.
```
Comment (CaseyHillers) [inline release_dashboard/lib/logic/git.dart:11]: This shouldn't be done in regex. Instead, rely on removing whitespace and splitting on the commas. Then you can loop over the parts against gitHashRegex

File: release_dashboard/lib/widgets/create_release_substeps.dart [hunk 20]
```diff
@@ -33,9 +34,35 @@ class CreateReleaseSubsteps extends StatefulWidget {
...
+  @visibleForTesting
+  Map<String, bool> isEachInputValid = <String, bool>{};
+
+  @override
+  void initState() {
+    // engine cherrypicks, framework cherrypicks and dart revision are optional and valid with empty input at the beginning
+    for (final String substep in CreateReleaseSubsteps.substepTitles) {
+      if (<String>[
```
Comment (CaseyHillers) [inline release_dashboard/lib/widgets/create_release_substeps.dart:48]: This would be cleaner by initializing the list of constants outside of the conditional

File: release_dashboard/lib/widgets/create_release_substeps.dart [hunk 21]
```diff
@@ -33,9 +34,35 @@ class CreateReleaseSubsteps extends StatefulWidget {
...
+        'Engine Cherrypicks (if necessary)',
+        'Framework Cherrypicks (if necessary)',
+        'Dart Revision (if necessary)'
+      ].contains(substep)) {
```
Comment (CaseyHillers) [inline release_dashboard/lib/widgets/create_release_substeps.dart:52]: This if check isn't necessary, you could do something like

```
        isEachInputValid = <String, bool>{
          ...isEachInputValid,
          substep: kOptionalValidators.contains(substep),
        };
```

File: release_dashboard/lib/widgets/create_release_substeps.dart [hunk 22]
```diff
@@ -112,30 +153,54 @@ class CreateReleaseSubstepsState extends State<CreateReleaseSubsteps> {
...
   final SetReleaseData setReleaseData;
   final String? hintText;
+  final ChangeIsEachInputValid changeIsInputValid;
 
   @override
   Widget build(BuildContext context) {
+    late RegExp formRegexValidator;
+    late String validatorErrorMsg;
+
+    Map<String, Object> RegexAndErrorMsg = git(name: substepName).getRegexAndErrorMsg();
```
Comment (CaseyHillers) [inline release_dashboard/lib/widgets/create_release_substeps.dart:178]: If we're only passing two values, it would be clearer to have a helper class like `GitValidator`, and have it offer `GitValidator.regex` and `GitValidator.hint`

Comment (CaseyHillers) [reply release_dashboard/lib/widgets/create_release_substeps.dart:178]: Generally, you want to avoid maps where possible when you know what the data is going to be.

File: release_dashboard/lib/widgets/create_release_substeps.dart [hunk 23]
```diff
@@ -112,30 +153,54 @@ class CreateReleaseSubstepsState extends State<CreateReleaseSubsteps> {
...
+      key: Key(substepName),
+      autovalidateMode: AutovalidateMode.onUserInteraction,
       decoration: InputDecoration(
-        labelText: CreateReleaseSubsteps.substepTitles[index],
+        labelText: substepName,
         hintText: hintText,
       ),
       onChanged: (String data) {
-        setReleaseData(CreateReleaseSubsteps.substepTitles[index], data);
+        data = data.trim();
```
Comment (CaseyHillers) [inline release_dashboard/lib/widgets/create_release_substeps.dart:190]: If users can enter multiple git hashes, we may want to do `data.replaceAll(' ', '')` to remove the whitespace (trim only removes from the ends)

File: release_dashboard/lib/widgets/create_release_substeps.dart [hunk 24]
```diff
@@ -145,27 +210,29 @@ class InputAsSubstep extends StatelessWidget {
...
     return Row(
       children: <Widget>[
         Text(
-          CreateReleaseSubsteps.substepTitles[index],
+          substepName,
           style: Theme.of(context).textTheme.subtitle1!.copyWith(color: Colors.grey[700]),
         ),
         // Only add a tooltip for the increment dropdown
-        if (index == 7)
+        if (substepName == CreateReleaseSubsteps.substepTitles[7])
```
Comment (CaseyHillers) [inline release_dashboard/lib/widgets/create_release_substeps.dart:235]: Avoid magic numbers. Try storing this as enum if it's required or having a more logical way to check for this (like a variable existing)

https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-mysterious-and-magical-numbers-that-lack-a-clear-derivation

File: release_dashboard/test/widgets/create_release_substeps_test.dart [hunk 25]
```diff
@@ -2,76 +2,281 @@
...
+          'Release Channel': true,
+          'Framework Mirror': true,
+          'Engine Mirror': true,
+          'Engine Cherrypicks (if necessary)': true,
+          'Framework Cherrypicks (if necessary)': true,
+          'Dart Revision (if necessary)': true,
+          'Increment': true,
+        }),
+      );
+    });
```
Comment (CaseyHillers) [inline release_dashboard/test/widgets/create_release_substeps_test.dart:280]: The tests lgtm!

File: release_dashboard/test/widgets/stepper_test.dart [hunk 26]
```diff
@@ -1,37 +0,0 @@
-// Copyright 2014 The Flutter Authors. All rights reserved.
-// Use of this source code is governed by a BSD-style license that can be
-// found in the LICENSE file.
-
-import 'package:conductor_ui/widgets/progression.dart';
-import 'package:flutter/material.dart';
-import 'package:flutter_test/flutter_test.dart';
-
-void main() {
-  testWidgets('When user clicks on a previously completed step, Stepper does not navigate back.',
```
Comment (CaseyHillers) [inline release_dashboard/test/widgets/stepper_test.dart:10]: Why was this removed?

File: release_dashboard/lib/logic/git.dart [hunk 27]
```diff
@@ -47,3 +47,63 @@ class git {
...
+
+/// Provides all the tools and methods to validate a multiple Git hash entry such as cherrypicks.
+class multiGitHash extends gitValidation {
+  /// Supports multiple git hashes delimited by a comma or an empty string.
+  ///
+  /// valid: c7714158950347,c7714158950347,c7714158950347
+  /// valid: c7714158950347
+  /// invalid:   c7714158950347,@@cccccc
+  /// invalid (cannot end with a comma):   c7714158950347,c7714158950347,
+  final RegExp multiGitHashRegex = RegExp(r'^$|^[0-9a-z]{40}(?:,[0-9a-z]{40})*$');
```
Comment (CaseyHillers) [inline release_dashboard/lib/logic/git.dart:85]: Instead of a multi git hash, we could simplify to `Iterable<bool> hashes = gitHashes.split(',').map((String gitHash) => GitHash.validate(githash));`.

Regexes are pretty much their own programming language, and should be avoided where possible.

File: release_dashboard/lib/logic/git.dart [hunk 28]
```diff
@@ -0,0 +1,121 @@
...
+  /// Returns the regex needed to validate this type of input.
+  RegExp get regex;
+
+  /// Returns the error message when the regex validation fails.
+  String get errorMsg;
+
+  /// Uses [regex] to validate the input provided. Returns true if it is valid, false otherwise.
+  ///
+  /// This method calls [sanitize] before the validation.
+  bool isValidate(String? input);
```
Comment (CaseyHillers) [inline release_dashboard/lib/logic/git.dart:17]: nit: usually the `ate` is dropped in favor of `isValid`

File: release_dashboard/lib/logic/git.dart [hunk 29]
```diff
@@ -0,0 +1,121 @@
...
+
+  /// Removes irrelevant characters such as whitespaces.
+  String sanitize(String? input);
```
Comment (CaseyHillers) [inline release_dashboard/lib/logic/git.dart:20]: If this is the same across all functions, why not just implement it here and inherit it in the children?

File: release_dashboard/lib/logic/git.dart [hunk 30]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  @override
+  bool isValidate(String? input) {
+    return regex.hasMatch(sanitize(input));
+  }
+}
+
+/// Provides all the tools and methods to validate a multiple Git hash entry such as cherrypicks.
+class MultiGitHash extends GitValidation {
```
Comment (CaseyHillers) [inline release_dashboard/lib/logic/git.dart:67]: Is this safe to remove now?

File: release_dashboard/lib/widgets/create_release_substeps.dart [hunk 31]
```diff
@@ -20,23 +32,44 @@ class CreateReleaseSubsteps extends StatefulWidget {
...
-    'Candidate Branch',
-    'Release Channel',
-    'Framework Mirror',
-    'Engine Mirror',
-    'Engine Cherrypicks (if necessary)',
-    'Framework Cherrypicks (if necessary)',
-    'Dart Revision (if necessary)',
-    'Increment',
-  ];
+  static Map<SubstepEnum, String> substepTitles = <SubstepEnum, String>{
```
Comment (CaseyHillers) [inline release_dashboard/lib/widgets/create_release_substeps.dart:35]: This is a great cleanup!

**Review by CaseyHillers — COMMENTED** [review]
Please rebase on tip of tree (there's a few analysis issues your lint pr will catch)

Other than that, I only have 2 small comments. Nice work!

**Review by CaseyHillers — APPROVED** [review]
flutter-lgtm
